Replace `pgxpool` with plain `pgconn` on PostgreSQL MCP server

gabrielcorado opened the pull request:
This PR migrates the PostgreSQL MCP server to use `pgconn` instead of `pgxpool` (`pgx`). A few reasons for this change:

- Follow the same flow and usage as in the PostgreSQL REPL (WebUI access).
- Better error propagation. Since `pgx` would wrap the execution inside transactions, some SQL errors weren't reported correctly, so we could not forward them back to the MCP client, making it harder to debug execution failures.
- Support for multiple queries on one execution. This follows the exact behavior of `psql` and WebUI REPL.
- Full transparency on what is executed on the database. Now, the query generated by the MCP client will be 1:1 on what is going to be executed on the database.
- Binary size. See the comparison below.

This change shouldn't change any behavior from the previous version (like closing connections after an inactivity period).

<details>
<summary>Binary size comparison</summary>

```
$ gsa tsh-master tsh-new
┌─────────────────────────────────────────────────────────────────────────────┐
│ Diff between tsh-master and tsh-new                                         │
├─────────┬───────────────────────────────────┬──────────┬──────────┬─────────┤
│ PERCENT │ NAME                              │ OLD SIZE │ NEW SIZE │ DIFF    │
├─────────┼───────────────────────────────────┼──────────┼──────────┼─────────┤
│ +0.05%  │ github.com/gravitational/teleport │ 28 MB    │ 28 MB    │ +13 kB  │
│ +0.15%  │ text/template                     │ 224 kB   │ 224 kB   │ +325 B  │
│ -0.00%  │ github.com/gogo/protobuf          │ 1.1 MB   │ 1.1 MB   │ -2 B    │
│ -0.65%  │ encoding/xml                      │ 104 kB   │ 104 kB   │ -676 B  │
│ -23.84% │ golang.org/x/sync                 │ 19 kB    │ 14 kB    │ -4.5 kB │
│ -1.82%  │ <autogenerated>                   │ 1.5 MB   │ 1.5 MB   │ -28 kB  │
│ -100%   │ github.com/jackc/puddle/v2        │ 44 kB    │          │ -44 kB  │
│ -89.38% │ github.com/jackc/pgx/v5           │ 3.1 MB   │ 330 kB   │ -2.8 MB │
├─────────┼───────────────────────────────────┼──────────┼──────────┼─────────┤
│ -0.28%  │ __go_buildinfo __DATA             │ 28 kB    │ 28 kB    │ -80 B   │
│ -0.01%  │ __noptrdata __DATA                │ 2.2 MB   │ 2.2 MB   │ -128 B  │
│ -3.62%  │ __itablink __DATA_CONST           │ 85 kB    │ 82 kB    │ -3.1 kB │
│ -3.31%  │ __typelink __DATA_CONST           │ 173 kB   │ 167 kB   │ -5.7 kB │
│ -1.64%  │ __data __DATA                     │ 818 kB   │ 805 kB   │ -13 kB  │
│ -1.83%  │ __rodata __TEXT                   │ 8.9 MB   │ 8.8 MB   │ -164 kB │
│ -2.88%  │ __rodata __DATA_CONST             │ 15 MB    │ 14 MB    │ -424 kB │
├─────────┼───────────────────────────────────┼──────────┼──────────┼─────────┤
│ -2.46%  │ tsh-master                        │ 119 MB   │ 116 MB   │ -2.9 MB │
│         │ tsh-new                           │          │          │         │
└─────────┴───────────────────────────────────┴──────────┴──────────┴─────────┘
```
</details>

gabrielcorado committed:
refactor(mcp): use raw pgconn on PostgreSQL MCP server

greedy52 commented on lib/client/db/postgres/mcp/mcp.go:
```diff
@@ -148,44 +220,71 @@ func (s *Server) wrapErrorResult(ctx context.Context, toolErr error) (*mcp.CallT
+		for _, row := range result.Rows {
+			rowData := make(map[string]string)
+			for columnIdx, contents := range row {
+				rowData[columns[columnIdx]] = string(contents)
```
how comfortable are we here just `string(contents)` without handling on individual types?

gabrielcorado replied:
We're using the PostgreSQL text format (the default when not set to binary), meaning that all messages are already converted into their text representation by the PostgreSQL server. This is the same mode used by `psql` and our REPL (which includes [a more extensive test](https://github.com/gravitational/teleport/blob/9ea168ba70bdc1b7ace895bf327f1ebd1ec5e4ee/lib/client/db/postgres/repl/testdata/TestStart/data_type.golden) covering the [most common types](https://github.com/gravitational/teleport/blob/9ea168ba70bdc1b7ace895bf327f1ebd1ec5e4ee/lib/client/db/postgres/repl/testdata/query.go#L29)).

Reference for PostgreSQL formatting: https://www.postgresql.org/docs/current/protocol-overview.html#PROTOCOL-FORMAT-CODES

I'll add this info to the code. Do you think we should have a similar test here? I've added it for REPL given that we do some formatting (basically fit into an ASCII table), but here since we're just returning as a plain value, I didn't see value in adding such a test.

greedy52 commented on lib/client/db/postgres/mcp/mcp.go:
```diff
@@ -252,10 +343,7 @@ const (
-	maxConnections = 1
+	// idleCheckerInterval is the interval which we'll check if the database
+	// connections are idle.
+	idleCheckerInterval = connectionIdleTime / 2
```
now looking at this again, is one minute idle timeout a little short?

greedy52 commented on lib/client/db/postgres/mcp/mcp_test.go:
```diff
@@ -169,57 +226,197 @@ func TestFormatErrors(t *testing.T) {
 
-func (mr *mockRows) CommandTag() pgconn.CommandTag {
-	return pgconn.NewCommandTag(mr.commandTag)
+func newTestDatabaseServer(t *testing.T) (chan *testDatabaseConn, dbmcp.DialContextFunc) {
```
i would prefer the traditional listener approach. took quite a while to understand what this does.

side talk, i know Krzysztof has looked into moving postgres test server to its own package but end up too big of a task. but maybe this can be a new starting point and we slowly migrate this version.

gabrielcorado replied:
> i would prefer the traditional listener approach.

Do you mean having a single goroutine processing all requests?

> side talk, i know Krzysztof has looked into moving postgres test server to its own package but end up too big of a task. but maybe this can be a new starting point and we slowly migrate this version.

I didn't use the test server to avoid additional usage, and since this use case is straightforward (we're only using it to test the idle checker, with no specific requirements for PostgreSQL).

I'll be honest, I don't like having a fake PostgreSQL server here as it feels we're retesting the `pgconn` library, but I'm still looking for a way to organize the code so we don't require it. Maybe moving the idle connection handler to rely solely on an interface for connection, and test it using dummy connections. I'll try some other alternatives here to see what I can get.

gabrielcorado committed:
chore(mcp): change godoc for postgres errors

gabrielcorado commented on lib/client/db/postgres/mcp/mcp_test.go:
```diff
@@ -169,57 +226,197 @@ func TestFormatErrors(t *testing.T) {
+		Databases:  []*dbmcp.Database{dbMCP},
+	})
+	require.NoError(t, err)
+	pgSrv := srv.(*Server)
```
We have the new MCP server defined as a shared type (so we can have a registry on the server), meaning we cannot change the return type (we could if we create another new function, but that feels unnecessary). But the `RunQuery` doesn't need to be public, I've renamed it.

https://github.com/gravitational/teleport/blob/f5cf64d7fd8f56b23263819e2ddf74adf2cb0f32/lib/client/db/mcp/mcp.go#L35-L36

gabrielcorado committed:
refactor(mcp): code review suggestions

gabrielcorado committed:
refactor(mcp): extract connections idler checker from pg

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/mcp-pg-move-to-pgconn

gabrielcorado committed:
chore(mcp): add missing godoc

gabrielcorado commented:
@greedy52 @GavinFrazar I've moved the idle checker to the `lib/client/db/mcp` package and made it reusable (for when we add other databases). It is a straightforward implementation and now has better tests for its usage, too. It also simplified the PostgreSQL tests as we don't require a test server anymore.

gabrielcorado committed:
refactor(mcp): wait until connections are released to close them

gabrielcorado committed:
test(mcp): add a close connections test case

greedy52 commented on lib/client/db/mcp/connections.go:
```diff
@@ -0,0 +1,230 @@
+
+	conn, ok := i.activeConnections[id]
+	if ok {
+		conn.acquire()
```
this will block `Get`. a query may take a long time then blocks other `id`?

gabrielcorado commented:
Closing in favor of #57590.
